### ISSUE TITLE
Hang / freeze when failing to push to Harbor oras registry

### ISSUE THREAD

**Issue by dcsouthwick** [sylabs/singularity#113]
**Version of Singularity**
```
$ singularity version
3.8.0
```
**Describe the bug**
Singularity 3.8.0 (both CE and forks) fail to catch & handle the error case where a .sif fails to push to Harbor.

sif images continue to be pushed using oras (see below ticket)

Relates to https://github.com/hpcng/singularity/issues/5691

**To Reproduce**

Have Harbor 2.2+ with OIDC (not username+password auth)

```
rm -rf $HOME/.singularity

singularity login -u <user> -p <harbor CLI secret> oras://your.harbor.instance

singularity -d push $HOME/alpine_latest.sif oras://your.harbor.instance/test/alpine:latest
DEBUG   [U=1000,P=18264]   persistentPreRun()            Singularity version: 3.8.0
DEBUG   [U=1000,P=18264]   persistentPreRun()            Parsing configuration file /usr/local/etc/singularity/singularity.conf
DEBUG   [U=1000,P=18264]   handleConfDir()               /home/dsouthwi/.singularity already exists. Not creating.
DEBUG   [U=1000,P=18264]   handleRemoteConf()            Ensuring file permission of 0600 on /home/dsouthwi/.singularity/remote.yaml
DEBUG   [U=1000,P=18264]   Init()                        Image format detection
DEBUG   [U=1000,P=18264]   Init()                        Check for sandbox image format
DEBUG   [U=1000,P=18264]   Init()                        sandbox format initializer returned: not a directory image
DEBUG   [U=1000,P=18264]   Init()                        Check for sif image format
DEBUG   [U=1000,P=18264]   Init()                        sif image format detected
DEBUG   [U=1000,P=18264]   UploadImage()                 ORAS push not accepted, retrying without config for registry compatibility

<wait 1-1000 minutes>
 
^CDEBUG   [U=1000,P=18264]   func2()                       User requested cancellation with interrupt
^C^C^C^C^C^C^C^C^C^C^C^C
```
program has froze/locked & unresponsive to interrupts. Reboot your session or machine.

**Expected behavior**
Ideally, pushing to oras harbor registry successfully.
If you do fail, at least timeout or something. Anything but locking up the session.

**OS / Linux Distribution**
Which Linux distribution are you using?
fails on centos/ubuntu/rhel in the same way. 

**Installation Method**
bug present for both source build and RPM (EPEL, fedora, etc)

**Comment by dtrudg** [comment]
Hi @dcsouthwick - I believe that we had this working previously as reported in #5691. We have changed to a new version of the ORAS library since then... that should be the only relevant change I believe. 

What version of the oras tool itself are you using that is working as expected?

**Comment by dcsouthwick** [comment]
For me, the latest (0.12) and 0.11.1 work, haven't gone back further. What version did you move to?

I will try and get some information on the registry side in the mean time

**Comment by dcsouthwick** [comment]
OK tracked down what seems to be going on here, and I don't think it's related to #5691.
Indeed, 3.8.0 pushes and pulls just fine to Harbor when everything is on a local filesystem, so the initial description above is inaccurate.
This hang/freeze is only exhibited when $HOME is on a nfs location (so far tested afs).

Steps to reproduce:
1. have a nfs/kerberized mounted $HOME
2. try pushing to harbor (potentially other oras//)

This is quite unexpected behavior to me, as singularity is supposedly run as my uid, which has rwx privs in my mounted nfs.
Also unexpected: setting $SINGULARITY_CACHEDIR to /tmp, pulling a .sif to /tmp, push _/tmp/alpine_latest.sif_ exhibits the same behavior. I guess the issue really stems from $HOME/.singularity/remote.yaml $HOME/.singularity/docker-config.json

Is there an ENV to place these two files somewhere else? I dont find anything obvious in the docs..

Sorry for the confusion. I hope there is a work-around for this, or ideally that I have simply missed something...

**Comment by dtrudg** [comment]
There isn't an env to change the `remote.yaml`. However, that shouldn't be the issue as we don't do anything other than a straight read of a file there, and presumably a `pull` from `library://` location is working? E.g. `singularity pull library://ubuntu`?

I'd expect it is more likely that something is getting hung up in the oras code... perhaps we are missing redirecting use of a path there.

Would you be able to provide debug output by running `singularity -d push oras://.....` and then output of `lsof` when it is stuck, showing anything relevant to singularity?

>This hang/freeze is only exhibited when $HOME is on a nfs location (so far tested afs).

I'm not clear here whether you have tested on NFS or AFS here? They are rather different filesystems, and may not exhibit the same issue. If you can provide the relevant lines from the output of `mount` that would be useful to confirm the fs and mount options etc.

**Comment by dcsouthwick** [comment]
Yes, pulling (both from a library:// and this harbor instance oras://) works just fine. 

Heres the initial lsof from a debug push to oras, captured while initially hung, then again after trying to kill (Ctrl+C) 

test was run on centos8 with afs as $HOME mount. I will try another filesystem if I get a bit of time.

Thanks,
/D

[lsof_singularity.log](https://github.com/sylabs/singularity/files/6748403/lsof_singularity.log)

**Comment by dtrudg** [comment]
Thanks, that's interesting as there is nothing showing in the lsof output for `$HOME`. I'll look into this further soon.

**Comment by dcsouthwick** [comment]
I suspect this is hidden within the sssd access that singularity is doing:
`/var/lib/sss/mc/passwd`
where my account is a network login (using sssd-krb5, sssd-ldap).
I'll try and dig more on my end

**Comment by dtrudg** [comment]
Singularity won't be interacting with sssd directly. It would only be doing so via the standard nss mechanism, that is going to be called through glibc. I don't see a path where this would be problematic specifically for an ORAS push, while other things work, but I could be wrong.

**Comment by dtrudg** [comment]
Hi @dcsouthwick - I'm afraid I'm struggling to reproduce this one, as it succeeds in my test environments using different network filesystems. At this point it might be useful to get `strace -f` output, tracing the execution of singularity and all threads/processes to see where it is getting stuck.

**Comment by vdechand** [comment]
Hello all,

this freeze happens on my workstation:
```
singularity version
3.8.1

Ubuntu 20.04.2 LTS

Harbor Version v2.3.1-1058f330 with authentication via OICD
```

```
singularity remote login -u USER oras://hub.url 
Password (or token when username is empty):  
INFO:    Token stored in /home/USER/.singularity/remote.yaml
```

```
singularity --debug push image.sif oras://hub.url/repo/image:latest
DEBUG   [U=1000,P=1165095] persistentPreRun()            Singularity version: 3.8.1
DEBUG   [U=1000,P=1165095] persistentPreRun()            Parsing configuration file /usr/local/etc/singularity/singularity.conf
DEBUG   [U=1000,P=1165095] handleConfDir()               /home/USER/.singularity already exists. Not creating.
DEBUG   [U=1000,P=1165095] handleRemoteConf()            Ensuring file permission of 0600 on /home/USER/.singularity/remote.yaml
DEBUG   [U=1000,P=1165095] Init()                        Image format detection
DEBUG   [U=1000,P=1165095] Init()                        Check for sandbox image format
DEBUG   [U=1000,P=1165095] Init()                        sandbox format initializer returned: not a directory image
DEBUG   [U=1000,P=1165095] Init()                        Check for sif image format
DEBUG   [U=1000,P=1165095] Init()                        sif image format detected
DEBUG   [U=1000,P=1165095] UploadImage()                 ORAS push not accepted, retrying without config for registry compatibility
^C
DEBUG   [U=1000,P=1165095] func2()                       User requested cancellation with interrupt
```

Not responding to SIGINT as well.

Same happens when pushing images to our organization's GitLab Container Registry, although "Root Image" appears there in the WebUI.

**Comment by dtrudg** [comment]
This hasn't been possible to replicate locally. Please could you provide strace output from `strace -f singularity --debug push image.sif oras://hub.url/repo/image:latest` so that we can see where it is getting stuck with more detail?

**Comment by vdechand** [comment]
Here is the strace output:
[strace_singularity.txt](https://github.com/sylabs/singularity/files/7168039/strace_singularity.txt)
